### PR TITLE
added ReadyReplicas status to notebook-controller

### PR DESCRIPTION
**Pull request by zabbasi**
#2723

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2743)

### CODE REVIEW

**Comment by zabbasi** [comment]
@lluunn

**Comment by vkoukis** [comment]
Hello @zabbasi,
This is in the right direction, and thanks for implementing it, but does not expose enough information to have compelling UX:

User:
> OK, my notebook server is not ready, but what is actually *doing*, is it stuck? what was its last error state, what can I do to fix this?

On the other hand, the UI can easily translate an ImagePullBackOff to a warning sign with a message similar to "Cannot download standard image `xyz` to start your Notebook Server. Please contact your administrator", or "Cannot download custom image `zyx` to start your Notebook Server. Please ensure you have typed its URL correctly", which is *much* more helpful.

I elaborate more on this in https://github.com/kubeflow/kubeflow/issues/2723#issuecomment-474666470.

I think we need to expose the actual status of the underlying pod, to enable this kind of UX.

File: components/notebook-controller/pkg/controller/notebook/notebook_controller.go [hunk 1]
```diff
@@ -103,11 +103,11 @@ type ReconcileNotebook struct {
 // and what is in the Notebook.Spec
 // Automatically generate RBAC rules to allow the Controller to read and write StatefulSet
 // +kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=apps,resources=statefulsets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=statefulsets/status,verbs=get;watch;update;patch
```
Comment (lluunn) [inline components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: I don't think you need this. and the one at line 110

File: components/notebook-controller/pkg/apis/notebook/v1alpha1/notebook_types.go [hunk 2]
```diff
@@ -39,6 +39,12 @@ type NotebookTemplateSpec struct {
 type NotebookStatus struct {
 	// Conditions is an array of current conditions
 	Conditions []NotebookCondition `json:"conditions"`
+
+	// readyReplicas is the number of Pods created by the StatefulSet controller that have a Ready Condition.
+	ReadyReplicas int32 `json:"readyReplicas"`
+
```
Comment (lluunn) [inline components/notebook-controller/pkg/apis/notebook/v1alpha1/notebook_types.go]: please remove blank line, also other places

File: components/notebook-controller/pkg/apis/notebook/v1alpha1/notebook_types.go [hunk 3]
```diff
@@ -39,6 +39,12 @@ type NotebookTemplateSpec struct {
 type NotebookStatus struct {
 	// Conditions is an array of current conditions
 	Conditions []NotebookCondition `json:"conditions"`
+
+	// readyReplicas is the number of Pods created by the StatefulSet controller that have a Ready Condition.
```
Comment (lluunn) [inline components/notebook-controller/pkg/apis/notebook/v1alpha1/notebook_types.go]: ready --> Ready

File: components/notebook-controller/pkg/controller/notebook/notebook_controller.go [hunk 4]
```diff
@@ -191,6 +192,17 @@ func (r *ReconcileNotebook) Reconcile(request reconcile.Request) (reconcile.Resu
 			return reconcile.Result{}, err
 		}
 	}
+	//Update the readyReplicas if the status is changed
```
Comment (lluunn) [inline components/notebook-controller/pkg/controller/notebook/notebook_controller.go]: // Update (add space)

**Comment by lluunn** [comment]
The test error is build-kfctl. You might need to merge master

**Comment by lluunn** [comment]
@vkoukis agreed.
But I think this PR is a good start. We can do the improvement in a followup PR?

**Comment by vkoukis** [comment]
@lluunn 
> @vkoukis agreed.
> But I think this PR is a good start. We can do the improvement in a followup PR?

Absolutely.
Thank you for acknowledging the importance having full visibility to the status of the underlying pod will have in e2e UX.
This PR is definitely a good start, looking forward to the followup PR, and we'll of course continue with #2588

**Comment by lluunn** [comment]
test error kfctl: #2760

**Comment by lluunn** [comment]
I don't see test failure..
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubeflow_kubeflow/2743/kubeflow-presubmit/6511/

/retest

**Pull request merged**